code/controllers/subsystem/spatial_gridmap.dm

```diff
@@ -0,0 +1,256 @@
+	for(var/datum/space_level/z_level as anything in SSmapping.z_list)
+		var/list/new_cell_grid = list()
+
+		grids_by_z_level += list(new_cell_grid)
+
+		for(var/y in 1 to cells_per_side)
+			new_cell_grid += list(list())
+			for(var/x in 1 to cells_per_side)
+				var/datum/spatial_grid_cell/cell = new(x, y, z_level.z_value)
+				new_cell_grid[y] += cell
+
+/**
+ * searches through the grid cells intersecting range radius around center and returns the added contents that are also in LOS
+ * much faster than iterating through view() to find all of what you want for things that arent that common
+ *
+ * * center - the atom that is the center of the searched circle
+ * * type - the grid contents channel you are looking for, see __DEFINES/spatial_grid.dm
+ * * range - the radius of our search circle. the code assumes this is > 1
+ * * ignore_visibility - if TRUE, line of sight is ignored, the contents of the grid are only filtered for distance
+ * * include_center - if FALSE, subtracts center from the output before filtering, used to speedup searches where you dont care about center being in the output
+ */
+/datum/controller/subsystem/spatial_grid/proc/find_grid_contents_in_view(atom/center, type, range, ignore_visibility = FALSE, include_center = TRUE)//should probably just be a global proc but w/e
+	var/turf/center_turf = get_turf(center)
+
+	var/center_x = center_turf.x
+	var/center_y = center_turf.y
+
+	var/list/contents_to_return = list()
+
+	var/static/grid_cells_per_axis = world.maxx / SPATIAL_GRID_CELLSIZE//im going to assume this doesnt change at runtime
+
+	//the minimum x and y cell indexes to test
+	var/min_x = max(CEILING((center_x - range) * (1 / SPATIAL_GRID_CELLSIZE), 1), 1)
+	var/min_y = max(CEILING((center_y - range) * (1 / SPATIAL_GRID_CELLSIZE), 1), 1)//calculating these indices only takes around 2 microseconds
+
+	//the maximum x and y cell indexes to test
+	var/max_x = min(CEILING((center_x + range) * (1 / SPATIAL_GRID_CELLSIZE), 1), grid_cells_per_axis)
+	var/max_y = min(CEILING((center_y + range) * (1 / SPATIAL_GRID_CELLSIZE), 1), grid_cells_per_axis)
+
+	var/list/grid_level = grids_by_z_level[center_turf.z]
+	switch(type)
+		if(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
+			for(var/list/grid_row as anything in grid_level.Copy(min_y, max_y + 1))//from testing, slightly faster than iterating numbers from min_index to max_index
+				for(var/datum/spatial_grid_cell/cell as anything in grid_row.Copy(min_x, max_x + 1))
+					contents_to_return += cell.client_contents
+
+		if(SPATIAL_GRID_CONTENTS_TYPE_HEARING)
+			for(var/list/grid_row as anything in grid_level.Copy(min_y, max_y + 1))
+				for(var/datum/spatial_grid_cell/cell as anything in grid_row.Copy(min_x, max_x + 1))
+					contents_to_return += cell.hearing_contents
+
+	if(!include_center) //this is faster for things that dont care about themselves but are (probably) in the output, helps us return without using the LOS algorithm
+		contents_to_return -= center
```

Comment:
Are we expecting to do a lot of -=ling/how large is the contents to return list?

Reply:
usually less than 10 elements as far as im aware. this check is so that you dont have to go through the line of sight algorithm if you dont get anything except yourself and you dont want to do anything with yourself (like radios)

Reply:
Aight. if we expected a large amount of items I'd tell you to make a new list of filtered things to return for the range/view loops, so we don't need to iterate over the list each time to remove an entry.

I wonder if iterating by index would be faster. that's too in the weeds though

Reply:
i thought that 
```
for(var/list/row as anything in grid.Copy(min_y, max_y + 1))
	for(var/datum/spatial_grid_cell/cell as anything in row.Copy(min_y, max_y + 1))
```
was faster from some old benchmarks, but in those i didnt realize that the last arg of Copy is one off from what i thought, so it was actually going through less cells. now it iterates through index which makes the proc ~10% faster. it also has to check if the list exists now which makes it ~4% slower (not an actual measurement), so it should be a little bit faster now and work with lazy content lists.